robocin/utility/README.md

```diff
@@ -0,0 +1,111 @@
+# utility
+
+A collection of utility functions and classes.
+
+## Table of Contents
+
+- [concepts](#concepts)
+- [epsilon](#epsilon)
+- [fuzzy_compare](#fuzzy_compare)
+- [type_traits](#type_traits)
+
+<a name="concepts"></a>
+
+## [`concepts`](concepts.h)
+
+The [concepts](concepts.h) header provides a set of concepts extended the standard library.
+
+- `arithmetic`: a concept that represents a type that is either an integral or floating point type;
+
+<a name="epsilon"></a>
+
+## [`epsilon`](epsilon.h.in)
+
+The [epsilon](epsilon.h.in) is a [CMake configure file](https://cmake.org/cmake/help/latest/command/configure_file.html)
+that defines the epsilon value passed to CMake during project build through the `ROBOCIN_FLOAT_EPSILON`,
+`ROBOCIN_DOUBLE_EPSILON` and `ROBOCIN_LONG_DOUBLE_EPSILON` flags (see [CMakeLists.txt](CMakeLists.txt)).
+
+These values will be used by the library to perform real number comparisons.
+
+* The existence of these values can be verified as follows:
+
+```cpp
+robocin::has_epsilon<float>::value;
+robocin::has_epsilon<double>::value;
+robocin::has_epsilon<long double>::value;
+```
+
+or:
+
+```cpp
+robocin::has_epsilon_v<float>;
+robocin::has_epsilon_v<double>;
+robocin::has_epsilon_v<long double>;
+```
+
+* When provided, you can access them without using:
+
+```cpp
+robocin::has_epsilon<float>::epsilon;
+robocin::has_epsilon<double>::epsilon;
+robocin::has_epsilon<long double>::epsilon;
+```
+
+or:
+
+```cpp
+robocin::epsilon_v<float>;
+robocin::epsilon_v<double>;
+robocin::epsilon_v<long double>;
+```
+
+<a name="fuzzy_compare"></a>
+
+## [`fuzzy_compare`](fuzzy_compare.h)
+
+The [fuzzy_compare](fuzzy_compare.h) header provides a set of functions to perform fuzzy comparisons and functor classes
+to be used with the standard library algorithms. There are many ways to compare real numbers, but for the applications
+that RobôCIn and other robotics teams use, the way chosen is to compare the absolute difference against
+a [tolerance margin](#epsilon):
+
+```cpp
+// check if a floating point number 'is zero':
+std::abs(value) <= epsilon;
+
+// check if two floating point numbers 'are equal':
+std::abs(lhs - rhs) <= epsilon;
+```
+
+- `fuzzyIsZero`: check if a floating point number is close to zero;
+- `fuzzyCmpEqual`: check if two floating point numbers are close to each other;
+- `fuzzyCmpNotEqual`: check if two floating point numbers are not close to each other;
+- `fuzzyCmpThreeWay`: check if two floating point numbers are close to each other, returning a C++20 three-way
```

Comment:
"check if" é algo meio aberto, não diz muito sobre o retorno da função, "returns true if a floating point number is close to zero" por exemplo é mais descritivo. Há uma falta de padrão no nome das funções, geralmente booleanas seguem com Is, assim como IsZero:
fuzzyIsEqual
fuzzyNotEqual
...

Reply:
Com exceção de `fuzzyIsZero`, todos os demais métodos possuíram inspiração na bibloteca padrão de comparação de numeros inteiros ([link](https://en.cppreference.com/w/cpp/utility/intcmp));

`isZero` trata-se, parcialmente, de uma exeção. Parcialmente por ser a unica função de aridade 1 presente nos comparativos, o que, na biblioteca padrão, faz-se uso do verbo to be `is` ([link](https://en.cppreference.com/w/cpp/utility/compare/named_comparison_functions)), dessa forma, seguiu-se esse padrão.

Em PRs futuros, parte da lógica adotada diferenciará metodos de funções, ex:
Caso tivesse sido optado o uso do verbo to be nas funções, teria sido preferido `areEqual`, porém, caso fosse um método, o primeiro se relacionaria com o segundo, fazendo sentido o uso de `isEqualTo`:
* `areEqual(lhs, rhs)` vs `lhs.isEqualTo(rhs)`;

Reply:
"check if" será atualizado para o sugerido.